src/main/java/org/mariadb/jdbc/internal/stream/PacketOutputStream.java

```diff
@@ -243,124 +257,138 @@ public boolean checkRewritableLength(int length) {
-                            buffer.position(buffer.position() + length);
-                        } else {
-                            bufferBytes = new byte[length];
-                            buffer.get(bufferBytes, 0, length);
-                            outputStream.write(bufferBytes, 0, length);
-                        }
-                        notCompressPosition += length;
-                        outputStream.flush();
-                    }
-                }
+        }
+    }
+    
+    private void flushWithCompression(int limit) throws IOException {
+        int notCompressPosition = 0;
+        int expectedPacketSize = limit + HEADER_LENGTH * ((limit / maxPacketSize) + 1);
+        byte[] bufferBytes = new byte[expectedPacketSize];
+        
+        while (notCompressPosition < expectedPacketSize) {
+            int length = buffer.remaining();
+            if (length > maxPacketSize) {
+                length = maxPacketSize;
             }
-
-
-            if (useCompression) {
-                //now bufferBytes in filled with uncompressed data
-                compressedAndSend(notCompressPosition, bufferBytes);
+            
+            int indexHeader = notCompressPosition;
+            bufferBytes[notCompressPosition++] = 0x00;
+            bufferBytes[notCompressPosition++] = 0x00;
+            bufferBytes[notCompressPosition++] = 0x00;
+            bufferBytes[notCompressPosition++] = 0x00;
+            
+            if (length > 0) {
+                int read = buffer.get(bufferBytes, notCompressPosition, length);
+                notCompressPosition += read;
+                bufferBytes[indexHeader++] = (byte) (read & 0xff);
+                bufferBytes[indexHeader++] = (byte) (read >>> 8);
+                bufferBytes[indexHeader++] = (byte) (read >>> 16);
+                bufferBytes[indexHeader++] = (byte) seqNo++;
+                
             }
         }
-
+        
+        // now bufferBytes in filled with uncompressed data
+        compressedAndSend(notCompressPosition, bufferBytes);
+    }
+    
+    private void flushRaw(int limit) throws IOException {
+        int expectedPacketSize = limit + HEADER_LENGTH * ((limit / maxPacketSize) + 1);
+        int notCompressPosition = 0;
+        
+        while (notCompressPosition < expectedPacketSize) {
+            int length = buffer.remaining();
+            if (length > maxPacketSize) {
+                length = maxPacketSize;
+            }
+            byte[] header = new byte[HEADER_LENGTH];
+            header[0] = (byte) (length & 0xff);
+            header[1] = (byte) (length >>> 8);
+            header[2] = (byte) (length >>> 16);
+            header[3] = (byte) seqNo++;
+            outputStream.write(header, 0, HEADER_LENGTH);
+            notCompressPosition += 4;
+            
+            if (length > 0) {
+                this.buffer.writeTo(this.outputStream);
```

Comment:
must write "length" bytes, like  this.buffer.writeTo(this.outputStream, 0, length);
Packet max size is set in variable maxPacketSize (server option max_allowed_packet). 
if sending more than this size, database will close socket.